### PR TITLE
Log error message when work branch is left behind

### PR DESCRIPTION
**Pull request by MilenaHristova**
https://github.com/dotnet/arcade/issues/12012

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/VmrOperationBase.cs [hunk 1]
```diff
@@ -117,9 +117,20 @@ SourceMapping ResolveMapping(string repo)
...
+                Logger.LogError("{message}", e.Message);
+                return false;
+            }
             catch (Exception e)
             {
-                Logger.LogError("Failed to synchronize repo {name}{exception}", mapping.Name, Environment.NewLine + e.Message);
+                Logger.LogError(
+                    "Failed to synchronize repo {name}{exception}." +
+                    Environment.NewLine +
+                    "A new branch was created for the sync and didn't get merged as the sync was interrupted. A new sync should start from the original branch.",
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/VmrOperationBase.cs:130]: This only applies for recursive synchronization/initialization actually.

Also, if we run the logic from somewhere else (MSBuild task) than Darc, we also want to see this.

For these reasons, I think you should move this catch block inside the `VmrInitializer`/`VmrUpdater` where we deal with the recursive processes.

I'd imagine you can catch, log this extra line and throw the exception further.

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrManagerBase.cs [hunk 2]
```diff
@@ -241,9 +241,19 @@ public static WorkBranch CreateWorkBranch(string repoPath, string branchName, IL
...
             using (var repo = new Repository(repoPath))
             {
-                logger.LogInformation("Creating a temporary work branch {branchName}", branchName);
-
                 originalBranch = repo.Head.FriendlyName;
+
+                if (originalBranch == branchName)
+                {
+                    var message = $"You are already on branch {branchName}. " +
+                                    "Previous sync probably failed and left the branch not merged. " +
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrManagerBase.cs:249]:
```suggestion
                                    "Previous sync probably failed and left the branch unmerged. " +
```

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/VmrOperationBase.cs (hunk 1 above)
Comment (premun) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/VmrOperationBase.cs:130]: Also, if you do it in the right place, you can tell user which branch is the original one (the recursive methods know this when they create the work branch).

Comment (MilenaHristova) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/VmrOperationBase.cs:130]: so it has to be caught inside the `UpdateRepositoryRecursively` and the `_currentBranch` field needs to be public

Comment (premun) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/VmrOperationBase.cs:130]: Possibly. Can it maybe be caught inside `MergeBack()`?

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrInitializer.cs [hunk 3]
```diff
@@ -90,19 +91,29 @@ public class VmrInitializer : VmrManagerBase, IVmrInitializer
...
+                    continue;
+                }
 
-            await InitializeRepository(update, cancellationToken);
+                await InitializeRepository(update, cancellationToken);
+            }
+        }
+        catch (Exception)
+        {
+            _logger.LogError("A new branch was created for the sync and didn't get merged as the sync " +
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrInitializer.cs:113]: Should this re-use the string? Seems the same in Updater?

Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrInitializer.cs:113]:
```suggestion
            _logger.LogWarning("A new branch was created for the sync and didn't get merged as the sync " +
```

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs [hunk 4]
```diff
@@ -330,7 +333,16 @@ public class VmrUpdater : VmrManagerBase, IVmrUpdater
...
 
-            var patchesToReapply = await UpdateRepositoryInternal(update, noSquash, false, cancellationToken);
+            IReadOnlyCollection<VmrIngestionPatch> patchesToReapply;
+            try
+            {
+                patchesToReapply = await UpdateRepositoryInternal(update, noSquash, false, cancellationToken);
+            }
+            catch(Exception)
+            {
+                _logger.LogError(interruptedSyncExceptionMessage, workBranch.OriginalBranch);
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs:343]:
```suggestion
                _logger.LogWarning(interruptedSyncExceptionMessage, workBranch.OriginalBranch);
```

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs [hunk 5]
```diff
@@ -358,7 +370,16 @@ public class VmrUpdater : VmrManagerBase, IVmrUpdater
...
         if (vmrPatchesToReapply.Any())
         {
-            await ReapplyVmrPatches(vmrPatchesToReapply.DistinctBy(p => p.Path).ToArray(), cancellationToken);
+            try
+            {
+                await ReapplyVmrPatches(vmrPatchesToReapply.DistinctBy(p => p.Path).ToArray(), cancellationToken);
+            }
+            catch (Exception)
+            {
+                _logger.LogError(interruptedSyncExceptionMessage, workBranch.OriginalBranch);
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs:379]:
```suggestion
                _logger.LogWarning(interruptedSyncExceptionMessage, workBranch.OriginalBranch);
```

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/VmrOperationBase.cs (hunk 1 above)
Comment (MilenaHristova) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/VmrOperationBase.cs]: there is a edge case - initializing a repo fails, current branch is `init/{mapping.Name}/..` then we run update, the original branch is the init branch, not main

Comment (premun) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/VirtualMonoRepo/VmrOperationBase.cs]: Maybe it's not possible, yeah

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs [hunk 6]
```diff
@@ -330,7 +330,19 @@ public class VmrUpdater : VmrManagerBase, IVmrUpdater
...
+            try
+            {
+                patchesToReapply = await UpdateRepositoryInternal(update, noSquash, false, cancellationToken);
+            }
+            catch(Exception)
+            {
+                _logger.LogWarning(
+                InterruptedSyncExceptionMessage,
+                workBranch.OriginalBranch.StartsWith("sync") || workBranch.OriginalBranch.StartsWith("init") ?
+                "the original" : workBranch.OriginalBranch);
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs:343]:
```suggestion
                    InterruptedSyncExceptionMessage,
                    workBranch.OriginalBranch.StartsWith("sync") || workBranch.OriginalBranch.StartsWith("init")
                        ? "the original"
                        : workBranch.OriginalBranch);
```

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs [hunk 7]
```diff
@@ -358,7 +370,19 @@ public class VmrUpdater : VmrManagerBase, IVmrUpdater
...
+            try
+            {
+                await ReapplyVmrPatches(vmrPatchesToReapply.DistinctBy(p => p.Path).ToArray(), cancellationToken);
+            }
+            catch (Exception)
+            {
+                _logger.LogWarning(
+                InterruptedSyncExceptionMessage,
+                workBranch.OriginalBranch.StartsWith("sync") || workBranch.OriginalBranch.StartsWith("init") ?
+                "the original" : workBranch.OriginalBranch);
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs:382]:
```suggestion
                    InterruptedSyncExceptionMessage,
                    workBranch.OriginalBranch.StartsWith("sync") || workBranch.OriginalBranch.StartsWith("init")
                        ? "the original"
                        : workBranch.OriginalBranch);
```

**Review by premun — COMMENTED** [review]
Just added small indentation fixes but otherwise LGTM.

I am just thinking whether we need this `WorkBranchException` at all? Maybe it's extra now?

**Comment by MilenaHristova** [comment]
> Just added small indentation fixes but otherwise LGTM.
> 
> I am just thinking whether we need this `WorkBranchException` at all? Maybe it's extra now?

yeah, I guess it's not needed now

**Review by premun — APPROVED** [review]

**Pull request merged**